### ISSUE TITLE
`x509-certificate` does not always return accurate fingerprint for `X509`

### ISSUE THREAD

**Issue by Hainish** [indygreg/PyOxidizer#525]
When running

```rust
x509.digest(MessageDigest::from_name("sha256").unwrap()).unwrap();
```

on a `openssl::x509::X509` object, the fingerprint returned is not always accurate.

I've updated the POC code in https://github.com/Hainish/parse-ber-poc to print the reported digests of the two RSA files.

Comparing this against 

```shell
keytool -printcert -file rsa-files/CIARANG.RSA 2>/dev/null | grep SHA256
keytool -printcert -file rsa-files/1.RSA 2>/dev/null | grep SHA256
```

We see that the `CIARANG.RSA` fingerprint is correct, but the `1.RSA` fingerprint does not match.

**Comment by indygreg** [comment]
Thanks for the report.

This probably has something to do with DER normalization.

Currently, `x509-certificate` DER encodes the parsed ASN.1 data structure and takes the digest of that.

OpenSSL calls `X509_digest()`, which calls `ASN1_item_digest()` on the passed `X509 *`. This calls `ASN1_item_i2d()`, which serializes some internal ASN.1 to DER.

So presumably the DER encoding of this crate isn't matching OpenSSL's.

**Comment by indygreg** [comment]
The issue appears to be with the `version` field again. On serialization, we're writing it out. However, since it wasn't present in the original BER certificate, we're introducing a primitive into the encoding and causing a fingerprint mismatch.

I should have realized this when making the parsing of the `version` field optional.

**Comment by indygreg** [comment]
I just released `x509-certificate` 0.9.0 and `cryptographic-message-syntax` 0.10.0 with a fix for this issue.
